Rework request parser

saralein opened the pull request:
- Divide into header and request line parsers
- Create request line object to transfer data

saralein committed:
Rework request parser

- Divide into header and request line parsers
- Create request line object to transfer data

saralein committed:
Add parameter handling

- Add parameter parser to request parsing
- Add parameter controller for route

zdavy commented on server/src/main/java/com/saralein/server/request/parser/RequestParser.java:
```diff
@@ -0,0 +1,55 @@
+
+    public Request parse(String rawRequest) throws Exception {
+        if (rawRequest.isEmpty()) {
+            throw new Exception("Bad request. Connection closed.");
```
🤔 Weird that parser has this domain knowledge about what the servers going to do with this exception.

saralein replied:
I think I see what you are saying.  You mean, parser knows it is a bad request.  Parser would have no clue that a connection would be closed.  Right?  Same will apply for the request line parser then.

zdavy replied:
Yeah, that's exactly it

zdavy commented on server/src/main/java/com/saralein/server/request/parser/RequestLineParser.java:
```diff
@@ -0,0 +1,33 @@
+import java.util.List;
+
+public class RequestLineParser {
+    public RequestLine parse(List<String> request) throws Exception {
```
I feel like a "parse" method insinuates the thing it's being passed will be parsed. There's an extra layer of logic here in the method body where it has to know which item in the collection needs to get parsed.

saralein replied:
OK, like it would only be passed the request line, header parser would only be passed the header lines.  This might help a bit with the weirdness of having `parseBody` use the header parser.  I can move some of that logic out to only send in what is needed to each.

saralein committed:
Update parsers to receive respective request lines

- Update exception message for parsers
- Modify to use double CRLF for splitting request

saralein committed:
Adjust request line parser to take only request line

damonkelley commented on server/src/main/java/com/saralein/server/request/parser/RequestParser.java:
```diff
@@ -0,0 +1,73 @@
+    }
+
+    private String pullBody(List<String> messageHeaderAndBody) {
+        if (messageHeaderAndBody.size() == 2) {
```
What does `2` mean?

damonkelley commented on server/src/main/java/com/saralein/server/request/parser/RequestParser.java:
```diff
@@ -0,0 +1,73 @@
+        String fullHeader = messageHeaderAndBody.get(0);
+        int crlfIndex = fullHeader.indexOf(CRLF);
+
+        if (crlfIndex != -1) {
```
What does this magical `-1` mean?

zdavy commented on server/src/main/java/com/saralein/server/request/parser/RequestParser.java:
```diff
@@ -0,0 +1,73 @@
+    }
+
+    private List<String> splitMessageHeaderAndBody(String request) {
+        return Arrays.asList(request.split(CRLF + CRLF));
```
does this mean the headers include the request line?

saralein replied:
Yeah, I was trying to use this HTTP language I found, but I think it's still unclear.  It divides the request into the message header and the body.  The message header contains the request line and the request headers.

saralein committed:
Adjust pull header to use contains check on CRLF

saralein committed:
Merge pull request #37 from saralein/parameter-parsing

Add parameter handling
